assayist/processor/base.py

```diff
@@ -139,19 +175,63 @@ def create_or_update_archive_artifact(self, archive_id, filename, arch, archive_
             _type = 'other'
         return self.__create_or_update_artifact(archive_id, _type, arch, filename, checksum)
 
-    def create_or_update_source_location(self, url, canonical_version):
+    def create_or_update_source_location(self, url, component, canonical_version=None):
         """
         Create or update SourceLocation.
 
+        Connect it to the component provided.
+        If canonical_version is provided link it with similar SourceLocations via SUPERSEDES.
+
         :param str url: the url and possibly commit hash of the source location
+        :param Component component: the component associated with this SourceLocation
         :param str canonical_version: (optional) the version of the component that
                                       corresponds with this commit id
         :return: a SourceLocation object
         :rtype: assayist.common.models.source.SourceLocation
         """
-        sl = source.SourceLocation.create_or_update({'url': url})[0]
+        # Figure out what type of SourceLocation we're dealing with.
+        sl_type = ''
+        if canonical_version:
+            # TODO: A better determination?
+            if '.redhat.com/' in url:
+                sl_type = 'local'
+            else:
+                sl_type = 'upstream'
+
+        # Get it.
+        sl = source.SourceLocation.create_or_update({
+            'url': url,
+            'type_': sl_type})[0]
+
         if canonical_version:
             sl.canonical_version = canonical_version
+
+            # Find all SLs related to this comonent of the same type.
+            # There is a match() function, but it only works if the relationship has a model.
+            similar_source_locations = component.source_locations
+            similar_source_locations = [x for x in similar_source_locations if x.type_ == sl_type]
+            similar_source_locations = [x for x in similar_source_locations if x.canonical_version]
+
+            # Find our place in the chain, which may be beginning, middle, or end.
+            # Note that I'm assuming that these are all in the chain. Which they should be.
+            similar_source_locations.append(sl)
+            if component.canonical_type == 'java':
+                key_method = generic_key
```

Comment:
To me it seems like it'd make sense to default to generic rather than RPM for version comparison. That way when we add additional content types, we don't have to keep adding conditionals to the if statement above.